components/places/src/storage/tags.rs

```diff
@@ -0,0 +1,221 @@
+
+use super::{fetch_page_info, TAG_LENGTH_MAX};
+use crate::error::{InvalidPlaceInfo, Result};
+use rusqlite::Connection;
+use sql_support::ConnExt;
+use url::Url;
+
+pub fn validate_tag(t: String) -> Result<String> {
+    // Drop empty and oversized tags.
+    let t = t.trim().to_string();
+    if t.len() == 0 || t.len() > TAG_LENGTH_MAX {
+        Err(InvalidPlaceInfo::InvalidTag(t).into())
+    } else {
+        Ok(t)
+    }
+}
+
+/// TODO: docstrings for all public functions.
+pub fn tag_url(conn: &Connection, url: &Url, tag: String) -> Result<()> {
+    let tag = validate_tag(tag)?;
+    let tx = conn.unchecked_transaction()?;
+
+    // This function will not create a new place.
+    // Fetch the place id, so we (a) avoid creating a new tag when we aren't
+    // going to reference it and (b) to avoid a sub-query.
+    let place_id = match fetch_page_info(conn, url)? {
+        Some(info) => info.page.row_id,
+        None => return Err(InvalidPlaceInfo::NoItem(url.to_string()).into()),
+    };
+
+    conn.execute_named_cached(
+        "INSERT OR IGNORE INTO moz_tags(tag, lastModified)
+         VALUES(:tag, now())",
+        &[(":tag", &tag)],
+    )?;
+
+    conn.execute_named_cached(
+        "INSERT OR IGNORE INTO moz_tags_relation(tag_id, place_id)
+         VALUES((SELECT id FROM moz_tags WHERE tag = :tag), :place_id)",
+        &[(":tag", &tag), (":place_id", &place_id)],
+    )?;
+    tx.commit()?;
+    Ok(())
+}
+
+pub fn untag_url(conn: &Connection, url: &Url, tag: String) -> Result<()> {
+    let tag = validate_tag(tag)?;
+    let tx = conn.unchecked_transaction()?;
+
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags_relation
+         WHERE tag_id = (SELECT id FROM moz_tags
+                         WHERE tag = :tag)
+         AND place_id = (SELECT id FROM moz_places
+                         WHERE url_hash = hash(:url)
+                         AND url = :url)",
+        &[(":tag", &tag), (":url", &url.as_str())],
+    )?;
+    // hrmph - do we actually need a transaction here?
+    tx.commit()?;
+    Ok(())
+}
+
+pub fn remove_all_tags_from_url(conn: &Connection, url: &Url) -> Result<()> {
+    let tx = conn.unchecked_transaction()?;
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags_relation
+         WHERE
+         place_id = (SELECT id FROM moz_places
+                     WHERE url_hash = hash(:url)
+                     AND url = :url)",
+        &[(":url", &url.as_str())],
+    )?;
+    // hrmph - do we actually need a transaction here?
+    tx.commit()?;
+    Ok(())
+}
+
+pub fn remove_tag(conn: &Connection, tag: String) -> Result<()> {
+    let tx = conn.unchecked_transaction()?;
+    conn.execute_named_cached(
+        "DELETE FROM moz_tags
+         WHERE tag = :tag",
+        &[(":tag", &tag)],
+    )?;
+    // hrmph - do we actually need a transaction here?
+    tx.commit()?;
+    Ok(())
+}
+
+pub fn get_urls_with_tag(conn: &Connection, tag: String) -> Result<Vec<Url>> {
```

Comment:
Should we order by frecency?

Reply:
I was thinking by last modified. Either (or no order) sounds fine to me!

Reply:
I went with frecency for `get_urls_with_tag()` under the assumption that a UI will order that way, but with lastModified for `get_tags_for_url()` - happy to adjust though.